### PR TITLE
http/s authonly transparent GET and POST with best effort json to map

### LINKED ISSUE #245: http interface with content-type based body parsing

### PR DESCRIPTION
**Pull request by c-bik**
fixes #245

### CODE REVIEW

File: src/imem_client.erl [hunk 1]
```diff
@@ -83,14 +86,41 @@ fix_git_raw_url(Url) ->
         nomatch -> error(bad_url)
     end.
 
-http_get(Url, Token) ->
-    case httpc:request(
-        get, {Url, [{"Authorization", "token " ++ Token}]},
-        [], [{body_format, binary}]
-    ) of
+http(Op, Url, ReqHeaders, Auth, Body) when is_map(Body) ->
```
Comment (shamis) [inline src/imem_client.erl]: I think it would be good to have a spec

File: src/imem_client.erl [hunk 2]
```diff
@@ -83,14 +86,41 @@ fix_git_raw_url(Url) ->
...
+http(Op, Url, ReqHeaders, Auth, Body) ->
+    http(Op, {Url, ReqHeaders, "application/text"}, Auth, Body).
+
+http(Op, {Url, ReqHeaders, ContentType}, {token, Token}, Body) ->
+    ReqHeaders1 = [{"Authorization","token " ++ Token} | ReqHeaders],
+    http(Op, {Url, ReqHeaders1, ContentType, Body});
+http(Op, {Url, ReqHeaders, ContentType}, {basic, User, Password}, Body) ->
+    Encoded = base64:encode_to_string(lists:append([User,":",Password])),
+    ReqHeaders1 = [{"Authorization","Basic " ++ Encoded} | ReqHeaders],
+    http(Op, {Url, ReqHeaders1, ContentType, Body}).
```
Comment (shamis) [inline src/imem_client.erl]: Request without authentication is not supported i think

File: src/imem_client.erl [hunk 3]
```diff
@@ -83,14 +86,41 @@ fix_git_raw_url(Url) ->
...
+              body => parse_http_resp(RespHeaders, RespBody)};
         {error, Error} -> error(Error)
     end.
+
+parse_http_resp([{_,_}|_] = RespHeaders, RespBody) ->
+    parse_http_resp(maps:from_list(RespHeaders), RespBody);
+parse_http_resp(#{"content-type" := ContentType}, Body) ->
+    case re:run(ContentType, "application/json") of
+        {match, _} ->
+            try imem_json:decode(Body, [return_maps]) catch _:_ -> Body end;
```
Comment (shamis) [inline src/imem_client.erl]: should this return with an error if the json is not parsable `{error, {invalid_json, Body}}` maybe

File: src/imem_client.erl [hunk 4]
```diff
@@ -83,14 +86,41 @@ fix_git_raw_url(Url) ->
...
               reasonPhrase => ReasonPhrase, headers => RespHeaders,
-              body => RespBody};
+              body => parse_http_resp(RespHeaders, RespBody)};
         {error, Error} -> error(Error)
     end.
+
+parse_http_resp([{_,_}|_] = RespHeaders, RespBody) ->
+    parse_http_resp(maps:from_list(RespHeaders), RespBody);
+parse_http_resp(#{"content-type" := ContentType}, Body) ->
+    case re:run(ContentType, "application/json") of
```
Comment (shamis) [inline src/imem_client.erl]: Should we match for `Content-Type` as well ?

File: src/imem_client.erl (hunk 2 above)
Comment (c-bik) [reply src/imem_client.erl]: For now there are no identified requirement!

File: src/imem_client.erl (hunk 4 above)
Comment (c-bik) [reply src/imem_client.erl]: changed to caseless match for content-type

File: src/imem_client.erl [hunk 5]
```diff
@@ -83,14 +86,64 @@ fix_git_raw_url(Url) ->
...
               reasonPhrase => ReasonPhrase, headers => RespHeaders,
-              body => RespBody};
+              body => parse_http_resp(RespHeaders, RespBody)};
         {error, Error} -> error(Error)
     end.
+
+parse_http_resp([{_,_}|_] = Headers, Body) ->
+    MaybeContentType =
+        lists:filtermap(
+            fun(Field, Value) ->
```
Comment (shamis) [inline src/imem_client.erl]: I think filtermap takes a function with 1 argument.

Comment (c-bik) [reply src/imem_client.erl]: Fixed! Please continue with review.

**Pull request merged**